### PR TITLE
GL_ENABLE_GET_PROC_ADDRESS

### LINKED ISSUE #20798: WebGL: extension string is available even though Emscripten doesn't implement it
Please include the following in your bug report:

**Version of emscripten/emsdk:**
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.50-git (9d830e22903a58a97e5aff79e31e05882112f8db)
clang version 18.0.0 (https://github.com/llvm/llvm-project e88a1ce0137bc14b754b86313228fdc48ba737f3)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /Users/kainino/sources/emsdk/upstream/bin
```

**Failing command line in full:**
`emcc main.cpp -o index.html -g -sUSE_WEBGL2`

**Full link command and output with `-v` appended:**
No build failure

Repro code: https://github.com/kainino0x/emscripten-extension-bug
Repro page: https://kai.graphics/emscripten-extension-bug/

Chrome Canary (as of 121.0.6153.0) recently enabled 4 new WebGL extensions:
EXT_clip_control
EXT_depth_clamp
EXT_polygon_offset_clamp
WEBGL_polygon_mode
https://chromiumdash.appspot.com/commit/40900b5e6698955aba32af9679519104c3bad20f

Emscripten passes through the extension strings from WebGL, but it does not actually implement the new entry points glClipControlEXT and glPolygonOffsetClampEXT. These will fail to link if you try to use them directly - but they'll return NULL if you try to get them with GetProcAddress, causing a crash in existing builds if they contain code to use these extensions:

```
EXT_polygon_offset_clamp was available, calling glPolygonOffsetClampEXT...
Uncaught RuntimeError: null function or function signature mismatch
    at index.wasm.__original_main (index.wasm:0x2dad)
    at index.wasm.main (index.wasm:0x2de3)
    at index.js:709:14
    at callMain (index.js:5683:15)
    at doRun (index.js:5733:23)
    at index.js:5744:7
```

To reproduce this make sure you have an updated Canary and EXT_polygon_offset_clamp is available on webglreport.com.
(If on Mac, you may need to switch to the Metal backend at `chrome://flags/#use-angle`.)
<details><summary>EDIT: I am going to revert these new extensions from Chrome per my analysis below. Hopefully a repro isn't really necessary but if needed, you can use an older Chromium snapshot:</summary>
<p>

Follow the Puppeteer instructions here:
https://www.chromium.org/getting-involved/download-chromium/#chrome-for-testing
Or use these snapshots:
https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Linux_x64/1229878/ (not tried)
https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Win_x64/1229878/ (not tried)
https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Mac_Arm/1229875/ (although I can't actually get this to work on my ARM mac)
https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Mac/1229878/ (nor these, though I thought they were dual-architecture binaries)

</p>
</details>

### PR DESCRIPTION
**Pull request by juj**
Some renderers that are ports from native codebases have a problem that they use the `*glGetProcAddress()` functions to dynamically discover the function entry points for different GL extensions.

In native GL world, there is a tacit guarantee that if an extension `GL_FOO` includes an entrypoint `glFoo()`, then if an implementation advertise support for `GL_FOO` extension, then calling `glGetProcAddress("glFoo")` can be expected to succeed.

More robust user code would check not only the presence of the extension, but also that the function `glGetProcAddress("glFoo")` does not return null, but such robustness is not necessarily built in to user applications since in the native GL world, this expectation doesn't generally fail.

Since some [6 years apo](https://github.com/emscripten-core/emscripten/commit/1facf3e687e7ede1c814f025683acfb6311f6f15#diff-aaacabb86431f19f9cc0196f449f8bc34764a5ac84556c21a7ce4934dddd9926R163), our best practices document has advocated users against using `*glGetProcAddress()` functions when targeting the web: https://emscripten.org/docs/optimizing/Optimizing-WebGL.html#optimizing-load-times-and-other-best-practices , due to the massive code size bloat that it causes (but also due to the performance problems that using it brings). Users are advised to not use that function.

Nevertheless, the function exists to ease porting existing codebases and to maintain parity between platforms.

One additional identified drawback is that this `*glGetProcAddress()` scheme enables GL code that is compiled to the web the possibility to activate "dormant" features from native GL side that were not enabled at the time the site is built; but later when a browser exposes a new WebGL extension, it can suddenly cause the site to try to activate those GL features.

This by itself wouldn't be a problem, except that **WebAssembly & WebGL do not support a facility to enable a true enumerating `*glGetProcAddress()` implementation**, but we must emulate that via a big if-else string search table. And so this statically implemented if-else chain cannot automatically manage new extension entry points as they come online when browsers evolve.

To fix above issues, this PR moves the `*glGetProcAddress()` functions behind a dedicated linker flag `-sGL_ENABLE_GET_PROC_ADDRESS=1` that users must now pass if they want to use a `*glGetProcAddress()` function. This enables our implementation to explicitly restrict to only advertising WebGL extensions that exist at the time that the site is built, so that the limited implementation of `*glGetProcAddress()` won't be able to break the promise that the existence of `GL_FOO` extension should imply that calling `glGetProcAddress("glFoo")` would succeed.

Fixes #20798

### CODE REVIEW

File: ChangeLog.md
Comment (sbc100) [inline ChangeLog.md:43]: Drop the `=1` here

File: emcc.py
Comment (sbc100) [inline emcc.py:940]: It should be enough to just check `USE_SDL` here since `tools/ports/sdl2_mixer.py` and `tools/ports/sdl2_gfx.py` both set this as part of `process_dependencies`

File: tools/system_libs.py
Comment (sbc100) [inline tools/system_libs.py:1833]: This doubles the number of libgl variants from 16 to 32.. which is a little unfortunate but maybe not avoidable?

File: tools/settings.py
Comment (sbc100) [inline tools/settings.py:85]: I wonder if this should be called `ENABLE_GL_GET_PROC_ADDRESS`?    Not sure it that is better or worse.

File: src/library_webgl.js
Comment (sbc100) [inline src/library_webgl.js:774]: We are still generally avoiding `let` and `const` in system libraries, for closure/code size reasons.

File: src/library_webgl.js
Comment (sbc100) [inline src/library_webgl.js:773]: Rather than interposing like this, can we instead explicitly call our wrapper 
function where it makes sense?   (then we can avoid this caching of the original function).

**Review by sbc100 — COMMENTED** [review]
There is quite a lot in the PR.  Perhaps we could split out the adding of support for the new extensions?  (that change could land either before or after the adding of the new setting).

File: ChangeLog.md
Comment (juj) [reply ChangeLog.md:43]: check

File: emcc.py
Comment (juj) [reply emcc.py:940]: check

File: tools/settings.py
Comment (juj) [reply tools/settings.py:85]: All GL settings have traditionally been prefixed in settings.py as `GL_xxx`. So I would not want to stray away from that convention. And `GL_ENABLE_GL_GET_PROC_ADDRESS` wouldn't look that good, hence condensing to just `GL_ENABLE_GET_PROC_ADDRESS`.

File: src/library_webgl.js
Comment (juj) [reply src/library_webgl.js]: check, changed to `var`

File: tools/system_libs.py
Comment (juj) [reply tools/system_libs.py:1833]: The variants system we have is really good at managing the combinatorial complexity, so this does not double our maintenance work. Looking at disk, the getProcAddress function is the bulk of the code size, and the non-getProcAddress enabled variants of the files are mostly just 2KB in size (with exception of that one offscreenframebuffer file), so I don't think the disk size increase amounts to anything.

File: src/library_webgl.js
Comment (juj) [reply src/library_webgl.js]: I think this is cleaner this way, as there will be just one code location where this is maintained. There are several places otherwise where this needs to be patched in, which would create dependencies across library_glemu.js, library_html5_webgl.js, library_webgl.js, webGLClient.js and webGLWorker.js.

**Comment by juj** [comment]
> There is quite a lot in the PR. Perhaps we could split out the adding of support for the new extensions? (that change could land either before or after the adding of the new setting).

I really prefer not to. I don't think the changes here have large technical complexity itself, it is mostly all just plumbing changes.

**Comment by juj** [comment]
The one thing to consider here is that this will be a breaking change, and possibly could be a breaking change to surprisingly many Emscripten users that build GLFW/GLEW/EGL based projects, since after this PR they will need to start passing `-sGL_ENABLE_GET_PROC_ADDRESS` in the link if they use `*glGetProcAddress()`. Currently the linker error comes from wasm-ld, so we don't have a mechanism to interject a helpful error message to the users. (maybe we could start intercepting wasm-ld if we care about this?)

It looks like SDL currently unconditionally uses that, so to mitigate the churn, I did make it enabled by default for SDL targets at least. I suppose I could do the same for GLFW/GLEW/EGL as well if we are concerned.

The recommendation for Filament would be to get rid of `*glGetProcAddress()` altogether for better codegen and GL performance.

File: system/include/webgl/webgl1_ext.h
Comment (kainino0x) [inline system/include/webgl/webgl1_ext.h:28]:
```suggestion
        #if EMSCRIPTEN_GL_WEBGL_polygon_mode
```

File: system/include/webgl/webgl1_ext.h
Comment (kainino0x) [inline system/include/webgl/webgl1_ext.h:64]: ?

File: system/include/webgl/webgl1_ext.h
Comment (kainino0x) [inline system/include/webgl/webgl1_ext.h:381]: ?

**Comment by sbc100** [comment]
How about if we land this new setting and start by only enabling in `-sSTRICT` mode?  (Which is for folks that are opting int these types of things early).

That would allows us to defer to decision around how and when to start breaking other users.

Regarding intercepting the wasm-ld error message and augmenting them.. that is something I've tried to avoid doing in the past and maybe we can avoid it here too: How about including a dummy version of `glGetProcAddress` in the JS library which triggers and meaningful error when included?  (see https://github.com/emscripten-core/emscripten/blob/main/src/library_exceptions_stub.js for an example of where we do this already)

**Review by kainino0x — COMMENTED** [review]
I didn't think about the fact that the code size would only need to impact GetProcAddress builds which are already quite large. Nice!

Note I've proposed renaming EXT_polygon_offset_clamp and EXT_clip_control to the WebGL WG so Emscripten will have to make changes if they do. This is fine though.

LGTM with the caveat that I'll leave decisions about breaking changes to other maintainers. I think it's a good change though.

**Comment by sbc100** [comment]
> > There is quite a lot in the PR. Perhaps we could split out the adding of support for the new extensions? (that change could land either before or after the adding of the new setting).
> 
> I really prefer not to. I don't think the changes here have large technical complexity itself, it is mostly all just plumbing changes.

That in kind of my point though.  All that unrelated plumbing boilerplate distracts from the functional stuff when trying to understand this PR.

I won't push hard for this, because I know how busy you likely are, but IMHO its always best to separate these things out, for the sake of the review process and also for future us who may be bisecting these changes.

File: system/include/webgl/webgl1_ext.h
Comment (juj) [reply system/include/webgl/webgl1_ext.h:28]: check

File: system/include/webgl/webgl1_ext.h
Comment (juj) [reply system/include/webgl/webgl1_ext.h:381]: Those functions were missing. I'll follow up with a separate PR to fix these two.

File: emcc.py
Comment (sbc100) [reply emcc.py:940]: I'm curious as to why we need all three here? Does this run before `process_dependencies`?

File: system/lib/gl/webgl2.c
Comment (sbc100) [inline system/lib/gl/webgl2.c:7]: Why is this outer `#if` required?

**Review by sbc100 — APPROVED** [review]
lgtm

BTW, what do you you think about putting this behind `-sSTRICT` first?   (then there would be no changes to the test/ files needed as part of this PR).  Maybe no need to such a period because we have good error reporting?

**Comment by juj** [comment]
> BTW, what do you you think about putting this behind -sSTRICT first?

My impression (given the "bad Emscripten" comments from linked thread) is that the WebGL people are regarding this to be a critical problem that prevents standardization from proceeding, so it doesn't seem like people should be given an opt-in choice.

I am happy that there will be a meaningful error message that hints how to fix the build error at least. Maybe there won't be too many angry users with the help of that.

I don't know if there are other users like Filament who might have an issue like this with their use of `*glGetProcAddress` and possibly dormant features that might be attempted to be accessed via `*glGetProcAddress`. As a data point, I checked that Unity has migrated away from using `*glGetProcAddress` a long time ago in August 2018.

**Comment by sbc100** [comment]
> (given the "bad Emscripten" comments from linked thread)

Sounds good to me.   

BTW what is '"bad Emscripten" comments from linked thread'?  (Sorry I don't see the link )

**Comment by juj** [comment]
> fix for "advertising extension we don't implement" [...] is completely separate and unrelated the "advertising extension we don't implement".

Maybe some typo?

> I just want to check my assumption that the bug fix and the new flag are separate?

No, the new flag is the bugfix.

Because there may exist codebases like Filament that make depend on that all enumerated GL extensions should be acquirable via `*glGetProcAddress()` function, we need to synchronize these two to go 1:1 hand in hand.

Utopistically, the correct fix to synchronize these two would be to repair `*glGetProcAddress()` to perform true enumeration of GL function pointers that WebGL extensions expose. But that is impossible to do in practical terms due to how WebAssembly-JS interop and WebGL extensions work.

So the realistic "fix" instead is to limit WebGL contexts to only advertise extensions that match 1:1 what `*glGetProcAddress()` is able to serve at the time that the codebase was compiled.

But that is a drastic pessimization in the big picture, and would be horrible to ship to all users, since only a small set of users utilize `*glGetProcAddress()`. Codebases that do not, and instead follow the best practices guide to statically link to WebGL extension functions are immune to this limitation of `*glGetProcAddress()`.

If we holistically applied this change to all codebases and not just those that use `*glGetProcAddress()`, we would have two major problems:
  a) the change would cause a code size impact also to developers that didn't suffer from the problem in the first place, and
  b) the change would needlessly require all users to update Emscripten whenever the want to "see" a new WebGL extension

By adopting this flag, we can target the above two limitations only to developers who actually do use `*glGetProcAddress()`. and other WebGL utilizing projects won't be disturbed.

File: src/library_webgl.js
Comment (sbc100) [inline src/library_webgl.js:745]: What about if the user calls `glGetString(GL_EXTENSIONS)` but they don't set `GL_ENABLE_GET_PROC_ADDRESS`?   Is it OK that they see the full unfiltered list in this case?

Similarly `emscripten_webgl_get_supported_extensions`.. if `GL_ENABLE_GET_PROC_ADDRESS` is not set what do we expect it to return?

How about GL_SUPPORT_AUTOMATIC_ENABLE_EXTENSIONS which is enabled by default.. do we want to see the full unfiltered list by default?

How about the existing wrapping of getSupportedExtensions by GL_DISABLE_HALF_FLOAT_EXTENSION_IF_BROKEN?  Does that wrapper simply layer in top of this one?

Comment (kainino0x) [reply src/library_webgl.js:745]: I think it should be OK to expose the unfiltered list in all cases when GetProcAddress is not enabled, because any code that uses an entry point Emscripten doesn't implement will cause a linker error. (And any extension that doesn't have entry points is OK.) It is a little unfortunate that they will have to `#ifdef __EMSCRIPTEN__` though instead of `#ifdef GL_EXT_clip_control` which many codebases will already have.

Come to think of it, it must be spec-valid in OpenGL for an implementation to list extensions that aren't available in the header - because you can use an older header with a newer driver. It's less guaranteed that applications would be robust to this though.

> GL_DISABLE_HALF_FLOAT_EXTENSION_IF_BROKEN

That wrapper shadows `WebGLRenderingContext.prototype.getSupportedExtensions` with `ctx.getSupportedExtensions` so I think it will apply first (at the JS level) and then this patch will apply on top.

Comment (kainino0x) [reply src/library_webgl.js:745]: > It's less guaranteed that applications would be robust to this though.

Er, my point is, Emscripten could remove them from the header instead of filtering them out, and that would be a less robust but technically valid fix (that would work for Filament at least in this case).

Actually it would be more robust than I thought - if we did this, we'd be disabling not just the static entrypoint `glClipControlEXT`, but also `PFNGLCLIPCONTROLEXTPROC` which should be used by any program that's using GetProcAddress on that function. Example: https://github.com/google/filament/blob/b007e9137eb029e256675eceb5fd52e6b873cecd/filament/backend/src/opengl/gl_headers.h#L132

**Comment by bkaradzic** [comment]
This breaks code building with new SDK, and once applied, it breaks build with old SDK.

I had to add `GL_ENABLE_GET_PROC_ADDRESS` to get GH actions to build with latest. But now when I build with previous version I get following error:

```
em++: error: Attempt to set a non-existent setting: 'GL_ENABLE_GET_PROC_ADDRESS'
```

**Comment by sbc100** [comment]
This seems like a generic problem that exists whenever we add a new setting to emscripten.   I support we could consider downgrading the `Attempt to set a non-existent setting` error to a warning so that it could be disabled? 

However, I wonder if this isn't a normal/common thing with all compiler tools.  When clang or gcc adds a new flag that flag isn't expected to be silently ignored by older versions of clang, is it?  I would assume the solution would be to detect which version of emcc is being used and set the new flag accordingly.. isn't that how new flags work in most command line applications?

**Comment by MarkCallow** [comment]
I was pointed to this PR as a victim of this change. My CI builds suddenly stopped working.

As a long time OpenGL user, the correct fix seems to me to be filtering the string returned by glGetExtensions to only include the extensions supported by the Emscripten OpenGL emulation. The other issue mentioned here, the code size when a glGetProcAddress function is used, seems a red herring. If the application is not calling a GetProcAddress function why is the code linked into it?

**Comment by sbc100** [comment]
> I was pointed to this PR as a victim of this change. My CI builds suddenly stopped working.

Can I ask how this change broke your build?   My understanding was that it should only have broken users who's programs reference the GetProcAddress family of function, but maybe I am misunderstanding?  What error are you seeing?

**Comment by MarkCallow** [comment]
> Can I ask how this change broke your build?

My code uses emscripten_GetProcAddress(). It uses *GetProcAddress because it is a part of a library which has many non-GL uses and I do not wish to force non-GL users to have to link with OpenGL.

In native OpenGL you have to (a) check if the implementation supports an extension by calling `glGetString` or `glGetStringi` and (b) call `*GetProcAddress` to retrieve any function pointers needed to use it.  ManyOpenGL drivers support multiple versions, as a result of which `*GetProcAddress` may return a non-NULL value even when the current context does not support the extension. Therefore any well-behaved OpenGL program wiil do (a) first.

Also the OpenGL spec. says

"GetString returns ... the extension names (in the EXTENSIONS string) that can be supported by the current GL context. Thus, if the client and server support different versions and/or extensions, a compatible ... list of extensions is returned."

We may consider WebGL to be the server in this case and Emscripten's emulation the client. Therefore filtering the extension list instead of returning WebGL's list verbatim will be GL spec. compliant and will fix the problem without a breaking change.

**Comment by juj** [comment]
Opened https://github.com/emscripten-core/emscripten/pull/20948 to flip the flag into opt-out from opt-in.

**Comment by kainino0x** [comment]
> The other issue mentioned here, the code size when a glGetProcAddress function is used, seems a red herring. If the application is not calling a GetProcAddress function why is the code linked into it?

It's not that it hurts code size for applications that don't use GetProcAddress - it's that it's a known "footgun" because GetProcAddress has a disproportionately huge code size, and when porting native applications it's easy to link it accidentally when you don't really need to.

**Pull request merged**
